File: donghwa/arxiv/topological_sort/802_find_safe_state.py
from collections import deque

def eventualSafeNodes(graph):
    n = len(graph)
    out_degree = [0] * n
    in_neighbors = [[] for _ in range(n)]
    
    for i, neighbors in enumerate(graph):
        out_degree[i] = len(neighbors)
        for neighbor in neighbors:
            in_neighbors[neighbor].append(i)
    
    queue = deque()
    for i in range(n):
        if out_degree[i] == 0:
            queue.append(i)
    
    safe_nodes = []
    while queue:
        node = queue.popleft()
        safe_nodes.append(node)
        for neighbor in in_neighbors[node]:
            out_degree[neighbor] -= 1
            if out_degree[neighbor] == 0:
                queue.append(neighbor)
    
    return sorted(safe_nodes)

# Test cases

graph1 = [[1, 2], [2, 3], [5], [0], [5], [], []]
result1 = eventualSafeNodes(graph1)
print(result1)
# assert result1 == [2, 4, 5, 6], f"Expected [2, 4, 5, 6] but got {result1}"

graph2 = [[1, 2, 3, 4], [1, 2], [3, 4], [0, 4], []]
result2 = eventualSafeNodes(graph2)
print(result2)
# assert result2 == [4], f"Expected [4] but got {result2}"

File: donghwa/arxiv/topological_sort/802_find_safe_state.cpp
/* 802. Find Eventual Safe States

There is a directed graph of n nodes with each node labeled from 0 to n - 1. The graph is represented by a 0-indexed 2D integer array graph where graph[i] is an integer array of nodes adjacent to node i, meaning there is an edge from node i to each node in graph[i].

A node is a terminal node if there are no outgoing edges. A node is a safe node if every possible path starting from that node leads to a terminal node (or another safe node).

Return an array containing all the safe nodes of the graph. The answer should be sorted in ascending order.

 

Example 1:

Illustration of graph
Input: graph = [[1,2],[2,3],[5],[0],[5],[],[]]
Output: [2,4,5,6]
Explanation: The given graph is shown above.
Nodes 5 and 6 are terminal nodes as there are no outgoing edges from either of them.
Every path starting at nodes 2, 4, 5, and 6 all lead to either node 5 or 6.
Example 2:

Input: graph = [[1,2,3,4],[1,2],[3,4],[0,4],[]]
Output: [4]
Explanation:
Only node 4 is a terminal node, and every path starting at node 4 leads to node 4.


*/
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
    int n = graph.size();
    vector<int> out_degree(n, 0);
    vector<vector<int>> in_neighbors(n);

    for (int i = 0; i < n; ++i) {
        out_degree[i] = graph[i].size();
        for (int neighbor : graph[i]) {
            in_neighbors[neighbor].push_back(i);
        }
    }

    queue<int> q;
    for (int i = 0; i < n; ++i) {
        if (out_degree[i] == 0) {
            q.push(i);
        }
    }

    vector<int> safe_nodes;
    while (!q.empty()) {
        int node = q.front();
        q.pop();
        safe_nodes.push_back(node);
        for (int neighbor : in_neighbors[node]) {
            out_degree[neighbor]--;
            if (out_degree[neighbor] == 0) {
                q.push(neighbor);
            }
        }
    }

    sort(safe_nodes.begin(), safe_nodes.end());
    return safe_nodes;
}


